doc/using/Foreign.rst

```diff
@@ -140,6 +143,9 @@ in C:
 
   extern int ghdl_main (int argc, char **argv);
 
+.. HINT::
```

Comment:
Strictly, it should not be an unused entry. Instead, it should be the path of the executed binary. Hence, I would not suggest to let it empty by default; we should explain two options:

- Pass/reuse the first argument that was passed to `main`, where `ghdl_main` is called from.
- Leave it empty/null.

Then, it is up to the user to decide.

---

Instead of adding a reference to a comment in some issue, I'd invite you to copy the relevant content from that comment to the docs. Actually, how to pass CLI arguments to GHDL might need to be clarified in a separate section, instead of being just a HINT. Relatively recent related issues: #1066, #819, #607.

Other meaningful references are https://ghdl.readthedocs.io/en/latest/using/InvokingGHDL.html#elaborate-and-run-elab-run (note `elab_options` and `run_options`) and https://ghdl.readthedocs.io/en/latest/using/Simulation.html (maybe "Simulation options" should be renamed to "Run options", to make it easier to understand that this is where `run_options` are defined).

So, the bare minimum rewrite would be to include references to those two sections in this HINT. From there on, up to the effort you want/can put on it.

Reply:
Sure. I have expanded on when to elaborate with the main.o. 

> maybe "Simulation options" should be renamed to "Run options"
I saw that Simulation and Runtime is the actual index. Even so, I was confused when trying to figure when I could use --disp-tree, namely that it did not work with --elab-run and it wasn't clear that it was a -r option. In the docs nor in the errors from ghdl. I think the examples we keep talking about would clear that up.

Reply:
> how to pass CLI arguments to GHDL might need to be clarified in a separate section

I left it in a hint. At most, it (CLI arguments via a custom main() to ghdl_main) could just be a sub under `Wrapping and starting a GHDL simulation from a foreign program`. I feel another revision coming on.

Reply:
> Sure. I have expanded on when to elaborate with the main.o.

I'm not sure about the location of the first hint you added... It feels like it would better fit as a rewrite/clarification of https://ghdl-rad.readthedocs.io/en/latest/using/Foreign.html#linking-foreign-object-files-to-ghdl. Then, https://ghdl-rad.readthedocs.io/en/latest/using/Foreign.html#wrapping-and-starting-a-ghdl-simulation-from-a-foreign-program would be specific about how to call `ghdl_main` explicitly. The build procedure remains the same as in the previous section. Should anyone jump here directly, you can add a first sentence suggesting the previous section to be read.

From there on, things start to get slightly complicated. The next section (https://ghdl-rad.readthedocs.io/en/latest/using/Foreign.html#linking-ghdl-to-ada-c) is about how to call GCC instead of `ghdl -e`. However, the "knowledge" of the two previous sections apply. Then, the build procedure that is required for https://ghdl-rad.readthedocs.io/en/latest/using/Foreign.html#dynamically-loading-ghdl is the same as in https://ghdl-rad.readthedocs.io/en/latest/using/Foreign.html#linking-ghdl-to-ada-c, but with `-shared`. In between, we find https://ghdl-rad.readthedocs.io/en/latest/using/Foreign.html#dynamically-loading-foreign-objects-from-ghdl, which might make sense to be moved below. To make thins more fun, *Linking foreign object files to GHDL*, *Wrapping and starting a GHDL simulation from a foreign program*, *Dynamically loading foreign objects from GHDL* and *Dynamically loading GHDL* can be used all at the same time. AFAIK, none of the methods are mutually exclusive.

Last, https://ghdl-rad.readthedocs.io/en/latest/using/Foreign.html#using-grt-from-ada is a kind of separate category. It is about interacting with the "runtime kernel". It would be interesting to extend it to C and/or Python, but I'm not aware of any work in progress in this regard.

Overall, we are not native english speakers, so any touch you can give to the page as a whole in order to make it easier to read will be really welcome. It does not need to be all in a single PR, tho. To have this merged, I think it is ok to just explain the first hint in the previous section.

> I saw that Simulation and Runtime is the actual index. Even so, I was confused when trying to figure when I could use --disp-tree, namely that it did not work with --elab-run and it wasn't clear that it was a -r option. In the docs nor in the errors from ghdl. I think the examples we keep talking about would clear that up.

> I left it in a hint. At most, it (CLI arguments via a custom main() to ghdl_main) could just be a sub under Wrapping and starting a GHDL simulation from a foreign program. I feel another revision coming on.

Let's discuss these in #1059, so we keep this PR simple.

Reply:
> To have this merged, I think it is ok to just explain the first hint in the previous section.

I did. I made it more generic too. It was a fair point you made. Thanks for your time